pkg/queue/breaker.go

```diff
@@ -99,25 +128,55 @@ type Semaphore struct {
 	mux      sync.Mutex
 }
 
-// Acquire receives the token from the semaphore, potentially blocking
+// Acquire receives the token from the semaphore, potentially blocking.
 func (s *Semaphore) Acquire() {
 	<-s.queue
 }
 
-// Release releases the token to the queue
-// The operation is potentially blocking when the queue is full
+// Release potentially puts the token back to the queue.
+// If the semaphore capacity was reduced in between and is not yet reflected,
+// we remove the tokens from the rotation instead of returning them back.
 func (s *Semaphore) Release() {
-	s.AddCapacity(1)
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.reducers > 0 {
+		s.capacity -= s.reducers
+		s.reducers--
+	} else {
+		// We want to make sure releasing a token is always non-blocking.
+		select {
+		case s.queue <- s.token:
+		default:
+		}
+	}
+}
+
+// AddCapacity increases the number of tokens in the rotation.
+// Increases only if the size is positive, does nothing otherwise.
+// An error is returned if not all capacity could be added to the semaphore.
+// This error could happen when the number of tokens exceeds the semaphore's queue size.
+func (s *Semaphore) AddCapacity(size int32) error {
+	if size > 0 {
```

Comment:
This if  is redundant, the loop will take care of that.

Reply:
yeap, will update